File: client/src/app/services/game-services/game-lobby.service.ts
import { Injectable } from '@angular/core';
import { GameManagementSocketEvent } from '@common/enums/socket-event/game-management-socket-event';
import { SocketFactoryService } from '@app/services/socket-service/socket-factory.service';
import { GameInfoService } from './game-info.service';
import { BehaviorSubject } from 'rxjs';
import { GameBaseService } from './base-classes/game-base.service';

@Injectable({
    providedIn: 'root',
})
export class GameLobbyService extends GameBaseService{
    isLobbyLocked: boolean = false;
    isFriendsOnly: boolean = false;
    canStartGame: BehaviorSubject<boolean>;

    constructor(
        socketFactoryService: SocketFactoryService,
        private gameInfoService: GameInfoService,
        
    ) {
        super(socketFactoryService);
        this.canStartGame ??= new BehaviorSubject(false);
        this.socketService.on(GameManagementSocketEvent.CanStartGame, (canStart: boolean) => this.canStartGame.next(canStart));
    }

    get gameId() {
        return this.gameInfoService.getGameId();
    }

    toggleLobbyLock(): void {
        this.socketService.emit(GameManagementSocketEvent.ToggleLock, undefined,
            (isLocked: boolean) => (this.isLobbyLocked = isLocked));
    }

    startGame() {
        this.socketService.emit(GameManagementSocketEvent.StartGame);
    }

    banPlayer(playerName: string) {
        this.socketService.emit(GameManagementSocketEvent.BanPlayer, playerName);
    }

    addBot() {
        this.socketService.emit(GameManagementSocketEvent.AddBot);
    }

    removeBot(playerName: string) {
        this.socketService.emit(GameManagementSocketEvent.RemoveBot, playerName);
    }

    updateBotDifficulty(playerName: string, difficulty: string) {
        this.socketService.emit(GameManagementSocketEvent.UpdateBotDifficulty, { playerName, difficulty });
    }

    resetState() {
        this.canStartGame.next(false);
        this.isLobbyLocked = false;
    }
}

File: android_app/lib/services/games/game_lobby_service.dart
import 'package:android_app/services/games/base_game_service.dart';
import 'package:android_app/services/sockets/socket_connection_service.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

final gameLobbyStateProvider =
    StateNotifierProvider<GameLobbyService, GameLobbyState>((ref) {
  return GameLobbyService(ref.read(socketConnectionServiceProvider));
});

final gameLobbyServiceProvider = Provider<GameLobbyService>((ref) {
  return ref.read(gameLobbyStateProvider.notifier);
});

class GameLobbyState {
  final bool isLobbyLocked;
  final bool canStartGame;

  const GameLobbyState({
    this.isLobbyLocked = false,
    this.canStartGame = false,
  });

  GameLobbyState copyWith({
    bool? isLobbyLocked,
    bool? canStartGame,
  }) {
    return GameLobbyState(
      isLobbyLocked: isLobbyLocked ?? this.isLobbyLocked,
      canStartGame: canStartGame ?? this.canStartGame,
    );
  }
}

class GameLobbyService extends BaseGameService<GameLobbyState> {
  GameLobbyService(SocketConnectionService socketService)
      : super(socketService, const GameLobbyState());

  void toggleLobbyLock() {
    socketService.emitWithAck('toggleLock', (isLocked) {
      state = state.copyWith(isLobbyLocked: isLocked);
    });
  }

  void addBot() {
    socketService.emit('addBot');
  }

  void updateBotDifficulty(String botName, String difficulty) {
    socketService.emit('updateBotDifficulty',
        {'playerName': botName, 'difficulty': difficulty});
  }

  void startGame() {
    socketService.emit('startGame');
  }

  void banPlayer(String playerName) {
    socketService.emit('banPlayer', playerName);
  }

  @override
  void initializeState() {
    state = const GameLobbyState();
  }

  @override
  void setUpSocketListener() {
    addSocketListener('canStartGame', (canStart) {
      state = state.copyWith(canStartGame: canStart);
    });
  }
}
